### PR TITLE
[REVIEW] IP Patch for NULL Count

### PR DESCRIPTION
**Pull request by BradReesWork**

### CODE REVIEW

File: include/gdf/cffi/types.h [hunk 1]
```diff
@@ -3,9 +3,9 @@
 typedef size_t gdf_size_type;
 typedef gdf_size_type gdf_index_type;
 typedef unsigned char gdf_valid_type;
-typedef	int64_t	gdf_date64;
-typedef	int32_t	gdf_date32;
-typedef	int32_t	gdf_category;
+typedef	long	gdf_date64;
```
Comment (jrhemstad) [inline include/gdf/cffi/types.h]: I assume these types need to be exposed to Pygdf, which is why they are in `types.h`?

File: src/io/csv/csv-reader.cu [hunk 2]
```diff
@@ -770,11 +787,12 @@ __global__ void convertCsvToGdf(
...
-				// // set the valid bitmap - all bits were set to 0 to start
-				int bitmapIdx 	= whichBitmap(rec_id);  	// which bitmap
-				int bitIdx		= whichBit(rec_id);		// which bit - over an 8-bit index
-				setBit(valid[col]+bitmapIdx, bitIdx);		// This is done with atomics
+			// set the valid bitmap - all bits were set to 0 to start
+			int bitmapIdx 	= whichBitmap(rec_id);  	// which bitmap
+			int bitIdx		= whichBit(rec_id);		// which bit - over an 8-bit index
+			setBit(valid[col]+bitmapIdx, bitIdx);		// This is done with atomics
 
+			atomicAdd((unsigned long long int*)&num_valid[col],(unsigned long long int)1);
```
Comment (jrhemstad) [inline src/io/csv/csv-reader.cu]: It looks like all threads are atomically incrementing a global variable. Consider using a shared memory counter per block, and then writing that to global memory at the end.

File: src/io/csv/csv-reader.cu [hunk 3]
```diff
@@ -422,40 +437,40 @@ gdf_error allocateGdfDataSpace(gdf_column *gdf) {
...
 	//--- allocate space for the valid bitmaps
-	CUDA_TRY(cudaMallocManaged(&gdf->valid, (sizeof(gdf_valid_type) 	* num_bitmaps)));
+	CUDA_TRY(cudaMalloc(&gdf->valid, (sizeof(gdf_valid_type) 	* num_bitmaps)));
 	CUDA_TRY(cudaMemset(gdf->valid, 0, (sizeof(gdf_valid_type) 	* num_bitmaps)) );
 
 	//--- Allocate space for the data
 	switch(gdf->dtype) {
 		case gdf_dtype::GDF_INT8:
-			CUDA_TRY(cudaMallocManaged(&gdf->data, (sizeof(int8_t) * N)));
+			CUDA_TRY(cudaMalloc(&gdf->data, (sizeof(int8_t) * N)));
```
Comment (jrhemstad) [inline src/io/csv/csv-reader.cu]: You can make your life a little easier here with `gdf_error get_column_byte_width(gdf_column * col, int * width)` which will tell you the byte width of the column's type. That way you only need 4 casts: 1, 2, 4, 8 bytes. https://github.com/gpuopenanalytics/libgdf/blob/master/src/column.cpp#L44

File: src/io/csv/csv-reader.cu (hunk 2 above)
Comment (BradReesWork) [reply src/io/csv/csv-reader.cu]: I agree there are some better ways to do this.  There are some upcoming tasks to break the code into primitives, null_count being one of them, and update for it is performed.

**Comment by scopatz** [comment]
LGTM!

**Pull request merged**